playwright/README.md

```diff
@@ -227,6 +227,107 @@ new AxeBuilder(page)
+
+The `FromFrames` object can be used as part of an existing `exclude` or `include` chain. The following shows how to test the `form` inside the `#payment` iframe, except for the `.ad-banner` in that `form`:
+
+```java
+import com.deque.html.axecore.args.FromFrames;
+
+new AxeBuilder()
+        .include(new FromFrames("iframe#payment", "form"))
+        .exclude(new FromFrames("iframe#payment", "form > .ad-banner"))
+```
+
+## Limit Shadow DOM Testing
+
+Including or excluding specific sections of a [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) tree can be done with a `FromShadowDom` selector object. This works similar to the [FromFrames](#limit-frame-testing) object selector.
+
+### AxeBuilder#include(FromShadowDom fromShadowDom)
+
+```java
+import com.deque.html.axecore.args.FromShadowDom;
+
+// Test each search form inside each <app-header> shadow DOM tree.
+new AxeBuilder()
+        .include(new FromShadowDom(".add-header", "form#search"))
+```
+
+The `FromShadowDom` selector object can also be used as part of an `exclude` or `include` method chain. It can be by itself, or with other selectors. The following example shows how to exclude all `.comment` elements inside the `<blog-comments>` custom element, as well as excluding the `footer` element:
+
+### AxeBuilder#exclude(FromShadowDom fromShadowDom)
+
+```java
+import com.deque.html.axecore.args.FromShadowDom;
+
+// Skip footer, as well as any .comment element inside the shadow DOM tree of <blog-comments>
+new AxeBuilder()
+        .exclude(".footer")
+        .exclude(new FromShadowDom("blog-comments", ".comment"))
+```
+
+The following shows how to test the `<app-footer>` custom component, inside the shadow DOM of the `#root` element, but to exclude any `.ad-banner` inside the `<app-footer>`'s shadow DOM tree:
+
+```java
+import com.deque.html.axecore.args.FromShadowDom;
+
+new AxeBuilder()
+        .include(new FromShadowDom("#root", "app-footer"))
+        .exclude(new FromShadowDom("#root", "app-footer", ".ad-banner"))
+```
+
+## Combine Shadow DOM and Frame Context
+
+To select frames inside shadow DOM trees or shadow DOM trees inside frames, it is possible to use [FromShadowDom](#limit-shadow-dom-testing) as a selector in the [FromFrames](#limit-frame-testing) selector object. The following example shows how to test the `main` element, inside each `iframe` that is part of the shadow DOM tree of `#appRoot`:
+
+```java
+import com.deque.html.axecore.args.FromFrames;
+import com.deque.html.axecore.args.FromShadowDom;
+
+new AxeBuilder()
+        .include(new FromFrames(new FromShadowDom("#appRoot", "iframe"), "main"))
+```
+
+The following shows how to exclude the `footer`, as well as any `.commentBody` elements in the `#userComments` shadow DOM tree, inside the `#blog-comments` iframe:
+
+```java
+import com.deque.html.axecore.args.FromFrames;
+import com.deque.html.axecore.args.FromShadowDom;
+
+new AxeBuilder()
+        .exclude("footer",
```

Comment:
Need to fix the formatting as right now it looks like `FromShadowDom` is a sibling to `FromFrames` rather than nested inside. 

Also, I believe the description is incorrect as this is an iframe selector, not two separate selectors. So this is excluding the `.commentBody` inside the `#userComments` shadow element, inside the `iframe#blog-comments` iframe, inside the `footer` iframe.